### ISSUE TITLE
expression node cloning isn't a deep copy

### ISSUE THREAD

**Issue by evykassirer** [josdejong/mathjs#745]
```
x = math.parse('2+2+3');
y = x.clone()
```

case 1:
```
x.args[1] = math.parse('555')
x.toString()
> "2 + 2 + 555"
y.toString()
> "2 + 2 + 3"
```

case 2:
```
x.args[0].args[0] = math.parse('56')
x.toString()
> "56 + 2 + 3"
y.toString()
> "56 + 2 + 3"
```

clone() is supposed to recursively copy (deep copy), right?

so this seems like a bug 😞

**Comment by FSMaxB** [comment]
The [code](https://github.com/josdejong/mathjs/blob/af551ff1cff769fe2979a30c60dd1224958960b2/lib/expression/node/Node.js#L218) suggests that it is supposed to be a shallow clone.

**Comment by evykassirer** [comment]
The documentation says it's a recursive clone, so one or the other should change.

**Comment by FSMaxB** [comment]
The documentation was changed [here](https://github.com/josdejong/mathjs/commit/72a1b0385078248d3e5888a69fd5978cc13907d7) and at the time it was recursive. It was changed to be a shallow clone in [this commit](https://github.com/josdejong/mathjs/commit/a1f60db0d15208a3be4e40813b59e9cd41c8d544). Not sure why.

**Comment by josdejong** [comment]
Hm. I agree this is counter intuitive. I guess I changed it because of performance reasons or something. 

Maybe we should change it to a deep copy again, with an optional flag to do a shallow clone (or a separate method).

**Comment by evykassirer** [comment]
found another issue with clone() - it doesn't preserve implicit multiplication in its cloning

```
x.toString() => 2 x
y = x.clone()
y.toString() => 2*x
```

Probably matters less to y'all, but it's important to the pedagogy of our math stepper.
For now, I'm accurately deep cloning by printing and reparsing, so this change isn't urgent, but it'd defs be nice :)

**Comment by FSMaxB** [comment]
Easy fix, see #747. Now I'm just waiting for Travis to run the tests.

**Comment by evykassirer** [comment]
Beautiful :) thanks

**Comment by josdejong** [comment]
I've implemented a new method `Node.cloneDeep`. 

So now we have a `Node.clone()` (shallow) and `Node.cloneDeep()` (deep), similar to libraries like lodash. 

Will do a release tonight.

**Comment by josdejong** [comment]
I've released v3.8.0 containing these fixes.
